src/Illuminate/Routing/RouteUrlGenerator.php

```diff
@@ -177,28 +168,16 @@ protected function formatParameters(Route $route, $parameters): array
             }
         }
 
-        // todo(wip) remove, just for debugging convenience.
-        // by this point we want to be sure we're working
-        // with a regular indexed array, so our array_pop() calls
-        // will not mess up anything related to named parameters
-        foreach (array_keys($parameters) as $k) {
-            assert(! is_string($k));
-        }
-
-        // note(wip) if the passed parameter count matches the count of
-        // parameters without defaults, we use the last passed parameter
-        // for the last route parameter without a default value and so on
-        // (handling cases where e..g a parameter with a default follows
-        // one without a default)
-        //
-        // if the passed count is lower, we already know we'll run into a
-        // "missing required parameter" situation. (note: if we used
-        // count($parameters) here instead the check WOULD have to be <=
-        // because we unset *named* parameters above)
-        //
-        // if the passed count is higher, we skip this step
-        if (count($passedParameters) == count($routeParametersWithoutDefaults)) {
-            foreach (array_reverse($routeParametersWithoutDefaults) as $name) {
+        // If the number of positional parameters matches the number
+        // of route parameters that don't have a default value or
+        // a named parameter, we prioritize matching positional
+        // parameters to these route parameters. This covers
+        // cases where a required parameter may be "wrapped"
+        // by parameters that have default values and are
+        // therefore optional, or their values have been
+        // provided as key-value pairs by the developer.
+        if (count($parameters) == count($routeParametersWithoutDefaultsOrNamedParameters)) {
```

Comment:
You could use `$passedParameterCount` here
```suggestion
        if ($passedParameterCount == count($routeParametersWithoutDefaultsOrNamedParameters)) {
```

Reply:
In this case it's using the count of `$parameters` *after* some values are removed. It isn't `$passedParameterCount` and switching to that makes tests fail.

Reply:
Oh, okay—thanks for the explanation 👍🏻  Nevermind

Reply:
It's why I back up the original count of the passed parameters. It might help to rename `$parameters` to something like `$remainingParameters` or `$positionalParameters` later to make this more clear.